### PR TITLE
Devcontainer / Github Codespace

### PR DESCRIPTION
**Pull request by colby-swandale**
## What was the end-user or developer problem that led to this PR?

Next week, contributors from the RubyGems/RubyGems.org/Bundler team will be attending the [RubyConf community day](https://rubyconf.org/faq#community-day), a day where attendees can meet face-to-face with Ruby Open-Source maintainers and learn the ins-and-outs of projects and possibly look to make some contributions.

To cut down on setup time and having maintainers fidgeting with various differing environments, I want to provide a way to  create environments super easily and have people onboarded with as little hassle as possible. 

## What is your fix for the problem, implemented in this PR?

This Pull Request adds [https://containers.dev](devcontainer ) / [GitHub Codespace](https://docs.github.com/en/codespaces/overview) configuration that automates:
* Setting up an isolated development environment
* Installs the necessary OS / Language / Library packages
* Setups the project to be ready for development.

After setup has been completed, users can then start writing code and testing their changes as if it were running on a MacBook / Linux Laptop.

## Notes

The devcontainer setup is fairly straightforward, but I made a trade-off with the Docker image the devcontainer is loaded into. Microsoft ships a [Ruby devcontainer](https://github.com/microsoft/vscode-dev-containers/tree/main/containers/ruby) image, which unfortunately uses the RVM tool to install and manage the Ruby Environment. We could manipulate the environment to not involve RVM, but I felt it was too confusing, and instead we should opt to use the well-known [base Ruby image](https://hub.docker.com/_/ruby/).

The trade-off for this means that we need to set up the devcontainer ourselves by:
* installing the required packages
* create a non-root user

But I'm confident this does not add much complexity that places a burden on maintainers and any issues couldn't be easily resolved.

### CODE REVIEW

**Comment by colby-swandale** [comment]
Anyone looking to test, you can create one from this branch via this link:

* https://github.com/codespaces/new?skip_quickstart=true&machine=basicLinux32gb&repo=614070&ref=colby%2Fdevcontainer&geo=UsWest&devcontainer_path=.devcontainer%2Fdevcontainer.json

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+{
```
Comment (simi) [inline .devcontainer/devcontainer.json:2]: Is it common to use this indentation in `devcontainer.json` file?

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+FROM ruby:${RUBY_VERSION}-alpine
+
+ENV USERNAME=vscode \
+    UID=1000 \
+    GID=1000
+
+# Install the git-credential-manager package via the dotnet tooling to 
+RUN apk update && apk add --no-cache \
+    github-cli \
```
[the remark this answers is not in the document]
Comment (colby-swandale) [reply .devcontainer/Dockerfile:11]: It's not strictly needed per se. The container will work fine with it, but it's handy to have around when working with Github Repositories & Codespaces. It doesn't feel harmful to include it by default.

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (colby-swandale) [reply .devcontainer/devcontainer.json:2]: This was generated from the VSCode Devcontainer extension, we can change it to match the rest of the RubyGems project.

Comment (colby-swandale) [reply .devcontainer/devcontainer.json:2]: Updated the indentation to be 2 spaces.

**Review by simi — APPROVED** [review]
Code-wise looks ok, but I have not tried. If you can run rubygems tests and random bundler spec in this container it is good to go from my side.

**Pull request merged**